results/pkg/api/server/server.go

```diff
@@ -155,54 +163,186 @@ func (s *Server) ListResultsResult(ctx context.Context, req *pb.ListResultsReque
+			pipelinerun pb.PipelineRun
+		)
+		if t := e.GetTaskRun(); t != nil {
+			taskrun = *(t)
+		}
+		if p := e.GetPipelineRun(); p != nil {
+			pipelinerun = *(p)
+		}
+		out, _, err := prg.Eval(map[string]interface{}{
+			"taskrun":     taskrun,
+			"pipelinerun": pipelinerun,
+		})
+		if err != nil && !strings.Contains(err.Error(), "no such attribute") && !strings.Contains(err.Error(), "undeclared reference to") {
+			log.Printf("failed to evaluate the expression: %v", err)
+			return false, status.Errorf(codes.InvalidArgument, "Error occurred during filter evaluation step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
+		}
+		if out.Value() == true {
+			return true, nil
+		}
 	}
+	return false, nil
+}
+
+// Encodes a result to PageToken
+func (s Server) encodeResult(primaryKey uint64) string {
+	var result []byte
+	if err := s.db.QueryRow("SELECT data from results where id=?", primaryKey).Scan(&result); err != nil {
+		return ""
+	}
+	encodedResult := make([]byte, base64.RawURLEncoding.EncodedLen(len(result)))
+	base64.RawURLEncoding.Encode(encodedResult, result)
+	return base64.RawURLEncoding.EncodeToString(encodedResult)
+}
 
-	// filter from all results
-	prg, err := s.env.Program(ast)
+func (s Server) decodeTokenToPrimaryKey(token string) (uint64, error) {
+	encodedToken, err := base64.RawURLEncoding.DecodeString(token)
+	if err != nil {
+		return 0, err
+	}
+	result := make([]byte, base64.RawURLEncoding.DecodedLen(len(encodedToken)))
+	_, err = base64.RawURLEncoding.Decode(result, encodedToken)
 	if err != nil {
-		log.Printf("program construction error: %s", err)
-		return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter checking step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
-	}
-	var resp []*pb.Result
-	for _, r := range results {
-		for _, e := range r.Executions {
-			out, _, err := prg.Eval(map[string]interface{}{
-				"taskrun": e.GetTaskRun(),
-			})
-			if err != nil {
-				log.Printf("failed to evaluate the expression: %v", err)
-				return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter evaluation step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
+		return 0, err
+	}
+	p := &pb.Result{}
+	if err := proto.Unmarshal(result, p); err != nil {
+		log.Printf("unmarshaling error: %v", err)
+		return 0, status.Errorf(codes.Internal, "failed to parse result data: %v", err)
+	}
+	var primaryKey uint64
+	if err := s.db.QueryRow("SELECT id FROM results where name=?", p.GetName()).Scan(&primaryKey); err != nil {
```

Comment:
This is susceptible to errors if the underlying result that was used for the pagination token is deleted, so we probably shouldn't depend on this.

If we create the internal proto as suggested above, we can just encode the primary key as part of the response and skip this step altogether.

Reply:
Oh got you, that's great~